### DISCUSSION TITLE
Fetching data inside component

### DISCUSSION THREAD

**Discussion by kaduppg** [vercel/next.js#28270] [Help] [answered]
Hey guys I've been working on a project using GraphQL and Craft CMS, and I've come across with this doubt when it comes to fetching data inside components instead of only use getStaticProps or getServerSideProps. 

The thing is ... Im pulling the home content blocks (data) using GraphQL and along with it the  Header and  Footer using getStaticProps, everything is working smoothly. After that I can pass down the props inside index.js, but the problem is I want to detach the Header and Footer components creating a BaseLayout component where I can wrap them and pass the children inside _app.js. 

I did this inside Header but I had to use useEffect to fetch the data using GraphQL, and it Worked. But I would like to see with you guys if this is a good practice or there's another approach that I can use , because as far as i know we can only use getStaticProps or getServerSideProps inside pages. 

Below you can see the structure:

```
Layout.js

import { Fragment } from "react";
import Footer from "./Footer";
import Header from "./Header";

const Layout = ({children}) => {
    return (
        <Fragment>
            <Header/>
                {children}
             <Footer/> 
        </Fragment>
    );
};

export default Layout;

```

_app.js

```
import Layout from '../components/Layout';

function MyApp({ Component, pageProps }) {

  return (
    <ApolloProvider client={client}>
     <Layout>
        <Component {...pageProps} />
     </Layout> 
    </ApolloProvider>
  )
}

export default MyApp


```

**Comment by delbaoliveira** [comment, marked as answer]
Hey @kaduppg. 

There are two main approaches on how you can fetch data in the client-side - with **[useSWR](https://swr.vercel.app/docs/data-fetching#graphql)** (a custom React hook created by the Vercel team) or with **useEffect**. 

With useSWR - You can take a look at the [example in our docs](https://nextjs.org/docs/basic-features/layouts#data-fetching) of how to fetch data for your layout component. You can point it to your API Route and use the Apollo Server to fetch your data, [here's an example](https://github.com/vercel/next.js/blob/canary/examples/api-routes-graphql/pages/api/graphql.js) from our repo.

With useEffect - Please take a look at [this article](https://www.apollographql.com/blog/apollo-client/next-js/next-js-getting-started/) from the Apollo team where they talk about how you can ensure the component only renders client-side.

**Reply by kaduppg** [reply]
Hey Delba, thanks a lot, Ill have a look and see which one will fits best \o/

**Reply by imran-baitham** [reply]
Hey Delba, thanks for help
